crates/vfio-ioctls/src/vfio_device.rs

```diff
@@ -388,12 +390,21 @@ impl VfioGroup {
     /// # Parameters
     /// * `id`: ID(index) of the VFIO group file.
     fn new(id: u32) -> Result<Self> {
-        let group_path = Path::new("/dev/vfio").join(id.to_string());
-        let group = OpenOptions::new()
-            .read(true)
-            .write(true)
-            .open(&group_path)
-            .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?;
+        let group = if cfg!(test) {
+            let tmp_file = TempFile::new().unwrap();
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(tmp_file.as_path())
+                .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?
+        } else {
+            let group_path = Path::new("/dev/vfio").join(id.to_string());
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(&group_path)
+                .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?
+        };
```

Comment:
I'm not a fan of introducing test specific code in the actual codepath that's being used for non-test cases. I'd rather have two functions `new()`, one for each case, even if this duplicates a little bit more code.